common/src/main/java/io/druid/collections/BlockingPool.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.collections;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Supplier;
+import com.metamx.common.ISE;
+import com.metamx.common.logger.Logger;
+
+import java.io.IOException;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Pool that pre-generates objects up to a limit, then permits possibly-blocking "take" operations.
+ */
+public class BlockingPool<T>
+{
+  private static final Logger log = new Logger(BlockingPool.class);
+
+  private final BlockingQueue<T> objects;
+
+  public BlockingPool(
+      Supplier<T> generator,
+      int limit
+  )
+  {
+    this.objects = limit > 0 ? new ArrayBlockingQueue<T>(limit) : null;
+
+    for (int i = 0; i < limit; i++) {
+      objects.add(generator.get());
+    }
+  }
+
+  public ResourceHolder<T> take() throws InterruptedException
+  {
+    Preconditions.checkState(objects != null, "Pool was initialized with limit = 0, there are no objects to take.");
+    return new ObjectResourceHolder(objects.take());
+  }
+
+  private class ObjectResourceHolder implements ResourceHolder<T>
```

Comment:
This part is similar, yeah. One difference is that this one adds back to the pool on finalize whereas stupid doesn't.